### PR TITLE
Certificate Generation

### PR DESCRIPTION
**Pull request by stevempope**
As specified in #64, the Leader of a Herd must be able to generate a certificate in order to encrypt map data sent out to members of the Herd.

My research has led me to the inbuilt Java security libraries. I went for these over using the OS certificates in the end as it keeps our program platform independent. The leader could in theory be any JVM compliant system.

This pull is in the scratch as I am not sure where we are placing leader/security in relation to each other.

I implemented the Singleton Pattern (GoF) in order to ensure that only one instance of the certificate can be created. We don't want self signed certificates just floating about the place!

This is alpha code, so it can be used to create a certificate and retrieve the public key. The next stage is to plug this into our Leader and RMI to distribute to all the members.

### CODE REVIEW

**Review by SamanthaDawsonBanks — APPROVED** [review]
code looks good, for clarity of people coming to the code code I question if following the singleton "make null then, if the null create and return else return" structure but that might be being too pedantic??

as for merging into the leader:
I would say leaver a separate class (aka solid) and maybe having that a leader has a cert in it (in same way a map has a gridded map in it)

**Comment by SamanthaDawsonBanks** [comment]
[aside] the decision to go with internal vs OS dependent cert gen was a good one.

As a team project it is important that all of use double check the suggestions we get for each other; if Steve had just blindly followed my suggestion then we would be OS locked (making development difficult) and coding towards a dead end.

great work Steve :)

File: mcomp_Scratchpad/src/security/Certificate.java [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+import java.security.KeyPairGenerator;
+import java.security.PrivateKey;
+import java.security.PublicKey;
+
+/**
+ * A class to generate and store a Public/Private Key pairing in
+ * the Java Keystore. Used by the Leader of a Herd to encrypt
+ * communication between itself and the Herd member.
+ * 
+ * This class may well merge with the Leader object, depending
```
Comment (RyanShoobert) [inline mcomp_Scratchpad/src/security/Certificate.java:13]: Agreed, we will have to look into the specifics of this at some point.

**Review by RyanShoobert — APPROVED** [review]
This is a good step towards getting security in the herd.

I agree with the idea of implementing in-built java libraires as this should mean our code will be able to run on a variety of devices (with JREs) and will be helpful for dev as well as running on the Raspberry Pis.

**Pull request merged**